Dev add i2s stm32h5

marwaiehm-st opened the pull request:
Add the STM32H5 family support for the test I2s_speed.
The driver is updated to handle correctly the commands TRIGGER_DRAIN.

francktheb committed:
dts: arm: st: h5: add I2S nodes

Addition of I2S nodes

Signed-off-by: Franck Thebault <franck.thebault@st.com>
Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

francktheb committed:
tests: drivers: i2s: speed: add stm32h5 support

Test on STM32H573i disco board , only the Tx transfer is ok.

Signed-off-by: Franck Thebault <franck.thebault@st.com>
Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

francktheb committed:
drivers: i2s : correct TRIGGER_DRAIN and TRIGGER_STOP commands

Correction of the handling of Tx audio samples via DMA

Signed-off-by: Franck Thebault <franck.thebault@st.com>
Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented:
@jeppenodgaard ^^

jeppenodgaard commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -597,12 +612,35 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/* check if all tx samples have been completely handled */
+		/*As stated in zephyr i2s specification, in case of DRAIN command */
+		/* Send all data in the transmit queue and stop the transmission.*/
```
I think multiline comment should follow linux' guidelines:
https://www.kernel.org/doc/html/latest/process/coding-style.html#commenting
This is also the case for other changes.

jeppenodgaard commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -597,12 +612,35 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+	/*if True, the tx stop command can be handled  */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/* check if all tx samples have been completely handled */
+		/*As stated in zephyr i2s specification, in case of DRAIN command */
```
Missing whitespace after * . Same in other changes.

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -491,6 +491,54 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
+				<&rcc STM32_SRC_PLL1_Q SPI1_SEL(0)>;
+			dmas = <&gpdma1 0 7 STM32_DMA_PERIPH_TX
+				&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_INC | \
```
`STM32_DMA_MEM_INC` is included in `STM32_DMA_PERIPH_RX`.

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -491,6 +491,54 @@
+				<&rcc STM32_SRC_PLL1_Q SPI1_SEL(0)>;
+			dmas = <&gpdma1 0 7 STM32_DMA_PERIPH_TX
+				&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
```
Should we use `STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS` in TX as well ?
If not, let's add a comment to explain this configuration.

Nit: A new flag `STM32_DMA_16BITS` could be added as a short to `STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS`

marwaiehm-st replied:
Yes , it should be used also in TX

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -491,6 +491,54 @@
+			reg = <0x40003800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00004000>,
+				<&rcc STM32_SRC_PLL1_Q SPI2_SEL(0)>;
+			dmas = <&gpdma1 2 9 (STM32_DMA_PERIPH_TX | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)
+				&gpdma1 3 8 STM32_DMA_PERIPH_RX>;
```
On i2s1 configuration is the opposite:
RX: 16 Bits, prio high

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -491,6 +491,54 @@
+			reg = <0x40003c00 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00008000>,
+				<&rcc STM32_SRC_PLL1_Q SPI3_SEL(0)>;
+			dmas = <&gpdma1 4 11 (STM32_DMA_PERIPH_TX  | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)
+				&gpdma1 5 10 STM32_DMA_PERIPH_RX>;
```
Ditto

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -491,6 +491,54 @@
+			dmas = <&gpdma1 0 7 STM32_DMA_PERIPH_TX
+				&gpdma1 1 6 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_INC | \
+					STM32_DMA_PERIPH_16BITS | STM32_DMA_MEM_16BITS | \
+					STM32_DMA_PRIORITY_HIGH)>;
```
Is that required only for RX ?

marwaiehm-st replied:
No, i added for the both (TX and RX)

erwango commented on tests/drivers/i2s/i2s_speed/boards/stm32h573i_dk.overlay:
```diff
@@ -0,0 +1,31 @@
+	status = "okay";
+};
+
+/* DMA Tx interrupt 0 has a higher Prio than Rx Dma interrupt 1 */
+&gpdma1 {
+	interrupts = <27 1 28 2 29 1 30 1 31 1 32 1 33 1 34 1>;
+	status = "okay";
+};
+
+&gpdma2 {
+	interrupts = <90 1 91 1 92 1 93 1 94 1 95 1 96 1 97 1>;
+	status = "okay";
+};
```
Is that required ?

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -174,9 +174,6 @@ static int i2s_stm32_configure(const struct device *dev, enum i2s_dir dir,
 	int ret;
 
 	if (dir == I2S_DIR_RX) {
-#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)
-		return -ENOSYS;
-#endif
```
Please make it a dedicated commit.

marwaiehm-st replied:
Done

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -652,11 +700,6 @@ static uint32_t i2s_stm32_irq_udr_count;
 static void i2s_stm32_isr(const struct device *dev)
 {
 	const struct i2s_stm32_cfg *cfg = dev->config;
-	struct i2s_stm32_data *const dev_data = dev->data;
-	struct stream *stream = &dev_data->rx;
-
-	LOG_ERR("%s: err=%d", __func__, (int)LL_I2S_ReadReg(cfg->i2s, SR));
-	stream->state = I2S_STATE_ERROR;
```
A dedicated commit would be required.

marwaiehm-st replied:
Done

marwaiehm-st committed:
include: zephyr: dt-bindings: dma: Add flag STM32_DMA_16BITS

Add flag STM32_DMA_16BITS as a combination of
STM32_DMA_PERIPH_16BITS and STM32_DMA_MEM_16BITS

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

mathieuchopstm commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -544,11 +544,9 @@ static int dma_stm32_reload(const struct device *dev, uint32_t id,
-		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id),
-				     size / stream->dst_size);
+		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id), size);
 	}
```
Identical code in `if` and `else` blocks; check should be removed.

marwaiehm-st replied:
I removed the check

marwaiehm-st committed:
drivers: i2s : enable rx path for H7 and H5

In i2s_stm32_isr, remove the errors from SR register
   i2s_stm32_configure, enable the rx path for H7 I2S compatible IP

Signed-off-by: Franck Thebault <franck.thebault@st.com>
Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st committed:
drivers: dma: u5: fix dma reload

In dma_stm32_reload, the size is the number of bytes and not number
of elements.

Signed-off-by: Franck Thebault <franck.thebault@st.com>
Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +349,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+				 * is triggered by DRAIN command
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else {
+				stream->stream_disable(stream, dev);
+				stream->queue_drop(stream);
+				stream->state = I2S_STATE_READY;
+			}
```
Non-RUNNING case already treated just above:
```
		if (stream->state != I2S_STATE_RUNNING) {
			irq_unlock(key);
			LOG_ERR("DRAIN trigger: invalid state");
			return -EIO;
		}
```

mathieuchopstm replied:
+1

The check here should probably be something like `if (/* stream queue not empty || stream DMA busy */)` instead.

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +349,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+		if (dir == I2S_DIR_TX) {
```
Should RX case be treated as documented: `	 * If the trigger is applied to the RX queue it has the same effect as * I2S_TRIGGER_STOP.` ?

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+				 */
+				stream->sample_tx_stop_drain = true;
+			} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+						(LL_SPI_IsActiveFlag_TXC(cfg->i2s))) {
```
An abstraction function will be required here to deal with F4/H5 variations:
```
static inline uint32_t ll_func_i2s_is_busy(I2S_TypeDef *i2s)
{
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)
	return LL_SPI_IsActiveFlag_TXC(i2s) == 0;
#else
	return LL_SPI_IsActiveFlag_BSY(i2s);
#endif /* st_stm32h7_i2s */
```

marwaiehm-st replied:
yes i forget to commit this modification :( i will do it now

marwaiehm-st committed:
drivers: i2s: Add DMA busy function

Add the function ll_func_i2s_dma_busy to be able to check the
transmission of all the DMA TX packet.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented on drivers/i2s/i2s_ll_stm32.h:
```diff
@@ -58,4 +61,16 @@ struct i2s_stm32_data {
 };
 
+/* checks that DMA Tx packet is fully transmitted over the I2S */
+static inline uint32_t ll_func_i2s_dma_busy(SPI_TypeDef *i2s)
```
```suggestion
static inline uint32_t ll_func_i2s_dma_busy(I2S_TypeDef *i2s)
```
?

marwaiehm-st replied:
The typdef of I2S does not exist in the hal_stm32, we used the spi type def in the structure i2s
```c
/* Device constant configuration parameters */
struct i2s_stm32_cfg {
	SPI_TypeDef *i2s;
	const struct stm32_pclken *pclken;
	size_t pclk_len;
	const struct pinctrl_dev_config *pcfg;
```

jeppenodgaard replied:
Why is this function not in the source file?

erwango commented on drivers/i2s/i2s_ll_stm32.h:
```diff
@@ -58,4 +61,16 @@ struct i2s_stm32_data {
+/* checks that DMA Tx packet is fully transmitted over the I2S */
+static inline uint32_t ll_func_i2s_dma_busy(SPI_TypeDef *i2s)
+{
+#ifdef LL_SPI_SR_TXC
```
#if #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_i2s)

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -30,6 +30,16 @@ static unsigned int div_round_closest(uint32_t dividend, uint32_t divisor)
 }
 
+static bool queue_is_empty(struct ring_buf *rb)
+{
```
This function is vulnerable to race conditions - while this isn't an issue ***now*** because all callers already suspend IRQs, it could become one in the future.
Please either:
* add a comment indicating this function must be called with IRQ suspended, or
* suspend interrupts for the duration of the function with an `irq_lock`/`irq_unlock` pair

Nit: this really should be a lock on the `struct ring_buf` instead of merely IRQ suspensions; this is acceptable(ish) because only one CPU at a time may interact with the DMA/I2S peripheral.

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+				stream->stream_disable(stream, dev);
+				stream->queue_drop(stream);
+				stream->state = I2S_STATE_READY;
+			} else
```
Please add braces around `else` block.

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +359,30 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+				return -EIO;
+		}
+
+		if (dir == I2S_DIR_RX) {
```
Please use `else if`.

Nit: replacing this code path with a `goto do_trigger_stop`, where `trigger_stop` is a label added in the `case I2S_TRIGGER_STOP` block above, could be cleaner, as it reduces code duplication and is more in line with the Zephyr API documentation:
```
If the [I2S_TRIGGER_DRAIN] trigger is applied to the RX queue it has the same effect as I2S_TRIGGER_STOP.
``` 

Nit 2: I2S_TRIGGER_STOP is currently broken - maybe a comment should be added to indicate this?

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -597,12 +639,41 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+		 * as stated in zephyr i2s specification, in case of DRAIN command
+		 * send all data in the transmit queue and stop the transmission.
+		 */
+		if (stream->sample_tx_block_sent == stream->sample_tx_block_dma_sent) {
```
Nit: what if you replace this with `queue_is_empty`?

marwaiehm-st replied:
I replace it with queue_is_empty , it didn't solve the last block problem but its more logical
 because the variable stream->sample_tx_block_sent it is initialized to 0 but never seems to be updated .

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.h:
```diff
@@ -41,6 +41,9 @@ struct stream {
 	uint8_t fifo_threshold;
+	uint8_t sample_tx_block_sent;
+	uint8_t sample_tx_block_dma_sent;
+	bool sample_tx_stop_drain;
```
Nit: variable names are confusing, maybe names such as the following could be more appropriate?
* `uint8_t blocks_queued_by_api`
* `uint8_t blocks_sent_by_dma`
* `bool tx_stop_for_drain`

mathieuchopstm reviewed (changes requested):
Mostly LGTM.
A few points to review:

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -336,10 +350,21 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+			stream->stream_disable(stream, dev);
+			stream->queue_drop(stream);
+			stream->state = I2S_STATE_READY;
+			stream->last_block = true;
```
This doesn't comply with the API description which states:
```
   Stop the transmission / reception of data at the end of the current
   memory block. This trigger can be used in RUNNING state only and at
   first changes the interface state to STOPPING. When the current TX /
   RX block is transmitted / received the state is changed to READY.
   Subsequent START trigger will resume transmission / reception where
   it stopped.
```

I think the whole `if`/`else if`/`else` can be replaced with something like this: (also fixes the logic bug above)
```c
if (/* dma busy */) {
  stream->state = STOPPING
  stream->tx_stop_for_drain = false;
} else {
  stream->stream_disable(/* ... */); //i.e., stop the DMA
  //update state, switch to READY
}
```

_Make sure **not** to call `stream->queue_drop()`._

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +374,29 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
-		stream->state = I2S_STATE_READY;
+
+		if (dir == I2S_DIR_TX) {
+			if (stream->state == I2S_STATE_RUNNING) {
```
Same logic bug as above. Similar pattern to above should work:

```c
if (/* dma busy */ || /* queue not empty */) {
   stream->state = STOPPING;
   stream->tx_stop_for_drain = true;
} else { //queue drained and no DMA op in progress
   //disable DMA
   //update state, switch to READY
}
```

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+			stream->queue_drop(stream);
+			stream->state = I2S_STATE_READY;
+			goto tx_disable;
+		} else {
```
Merge nested `else`+`if` into `else if` for improved readability.
Also add a comment like:
```c
/* else: DRAIN trigger -> continue TX normally until queue is empty */
```

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+	/*
+	 * Check if the tx stop command can be handled
+	 * if True, the tx stop command can be handled
+	 */
```
Nit: outdated/unclear comment?
```c
/* Check if we finished transferring one block and stopping is requested */
```

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -529,6 +574,16 @@ static void dma_rx_callback(const struct device *dma_dev, void *arg,
+			/*
+			 * In case of STOP command, just stop the transmission
+			 * at the current. The transmission can be resumed
+			 */
```
Comment refers to `transmission` but this is RX path.

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -529,6 +574,16 @@ static void dma_rx_callback(const struct device *dma_dev, void *arg,
 	mblk_tmp = stream->mem_block;
 
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		if (stream->tx_stop_for_drain == false) {
```
There is no need to check this on RX because `I2S_TRIGGER_DRAIN` is redirected to `I2S_TRIGGER_STOP`. Ignoring it also removes the need to update it in `i2s_stm32_trigger`.

mathieuchopstm reviewed (changes requested):
We're getting there! A few points to iron out:

ajarmouni-st commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -603,6 +659,33 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
+		 */
+		if (queue_is_empty(&stream->mem_block_queue) == true) {
+			LOG_INF("All the data has been transferred");
+			stream->queue_drop(stream);
```
Also `LOG_INF` here should be `LOG_DBG`

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +371,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
+			 */
+			stream->tx_stop_for_drain = true;
+		} else {
+			/* else: DRAIN trigger -> continue TX normally until queue is empty */
```
misplaced comment

mathieuchopstm commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -349,9 +371,20 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
+
```
```c
if (dir == I2S_DIR_RX)
  goto trigger_stop;
```
is missing again

mathieuchopstm reviewed (approved):
Tiny non-blockers, otherwise LGTM.

Nit: `stream->last_block` seems to have become a dead variable, could used to be removed if so.

marwaiehm-st committed:
drivers: i2s: i2s_ll_stm32: tx_stream_disable: Add busy wait

Add a k_busy_wait to the function tx_stream_disable
to be able to complete the drain of the TX queue before disabling.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

jeppenodgaard commented on drivers/i2s/i2s_ll_stm32.h:
```diff
@@ -58,4 +59,16 @@ struct i2s_stm32_data {
 };
 
+/* checks that DMA Tx packet is fully transmitted over the I2S */
+static inline uint32_t ll_func_i2s_dma_busy(SPI_TypeDef *i2s)
```
Why is this function not in the source file?

erwango replied:
We're already doing this in spi driver for function which are only providing hw abstraction

erwango replied:
Improving driver file clarity. I agree that in current case, for one function, this might appear an excessive measure.
For spi driver case, there are lot of functions abstracted so it really made sense (this is all subjective, I agree).
Reusing the same scheme is for coherency sake across STM32 drivers.

mathieuchopstm reviewed (approved):
LGTM.
